llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -4734,6 +4734,89 @@ void InstCombinerImpl::tryToSinkInstructionDbgValues(
   }
 }
 
+// If we have:
+//  `(op X, (zext/sext (icmp eq X, C)))`
+// We can transform it to:
+//  `(select (icmp eq X, C), (op C, (zext/sext 1)), (op X, 0))`
+// We do so if the `zext/sext` is one use and `(op X, 0)` simplifies.
+Value *InstCombinerImpl::foldOpOfXWithXEqC(Value *Op, const SimplifyQuery &SQ) {
```

Comment:
I'd like to handle `add X, (zext/sext (icmp eq/ne X, C))` first. The current version looks like an over-generalization.

Reply:
If I remember correctly, similar patterns like `X * (X != 0) -> X` exist already. Can they be folded here too?

Reply:
The `X != C` case isn't the same. We can handle if if `(op x, 1)` simplifies which is a lot rarer than `(op x, 0)`.

Reply:
> I'd like to handle `add X, (zext/sext (icmp eq/ne X, C))` first. The current version looks like an over-generalization.

I can see the argument that the intrinsics complicate the situation a bit and can drop those if you feel strongly, but imo the rest of the code is basically the same if its just add vs any op.

Reply:
FWIW, I think the additional complexity is actually quite significant. Doing this for just add is something like:

```
  if (match(I, m_c_Add(m_Value(X),
                       m_ZExt(m_CombineAnd(
                           m_Value(Cond),
                           m_ICmp(Pred, m_Deferred(X), m_ImmConstant(C)))))) &&
      Pred == ICmpInst::ICMP_EQ) 
    return Builder.CreateSelect(
        Cond, Builder.CreateAdd(C, ConstantInt::get(C->getType(), 1)), X);
```

Generalizing it to other operations is a hundred lines of code here, plus a dozen call sites elsewhere.

Reply:
Okay, let me test which operators get hit. If its just `add` ill drop all this, otherwise think its worth the generlization.

Reply:
Okay, created https://github.com/llvm/llvm-project/pull/93840